crates/iota-indexer/src/backfill/runner.rs

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    cmp,
+    collections::BTreeSet,
+    ops::RangeInclusive,
+    sync::{
+        Arc,
+        atomic::{AtomicUsize, Ordering},
+    },
+    time::Instant,
+};
+
+use futures::{StreamExt, TryStreamExt, stream::unfold};
+use tokio::sync::Mutex;
+use tokio_stream::Stream;
+use tracing::{error, info};
+
+use crate::{
+    backfill::{Backfill, BackfillKind, get_backfill},
+    config::BackfillConfig,
+    db::ConnectionPool,
+    errors::IndexerError,
+};
+
+/// Entry point for orchestrating backfills.
+///
+/// `BackfillRunner` selects the appropriate backfill implementation, splits the
+/// requested checkpoint range into manageable chunks, and dispatches them in
+/// parallel.
+pub struct BackfillRunner;
+
+impl BackfillRunner {
+    /// Execute a backfill over `total_range` using the specified task kind.
+    pub async fn run(
+        runner_kind: BackfillKind,
+        pool: ConnectionPool,
+        backfill_config: BackfillConfig,
+        total_range: RangeInclusive<usize>,
+    ) -> Result<(), IndexerError> {
+        let backfill = get_backfill(runner_kind, *total_range.start()).await;
+        Self::run_impl(pool, backfill_config, total_range, backfill).await
+    }
+
+    async fn run_impl(
+        pool: ConnectionPool,
+        config: BackfillConfig,
+        total_range: RangeInclusive<usize>,
+        backfill: Arc<dyn Backfill>,
+    ) -> Result<(), IndexerError> {
+        let timer = Instant::now();
+        let processed_counter = Arc::new(AtomicUsize::new(0));
+        // Keeps track of the ranges (using starting range number)
+        // that are in progress.
+        let in_progress = Arc::new(Mutex::new(BTreeSet::new()));
+
+        // Generate chunks
+        let chunk_stream = chunk_range_stream(total_range, config.chunk_size);
+
+        // Process chunks in parallel, fail-fast on error
+        chunk_stream
+            .map(|range| {
+                let pool = pool.clone();
+                let backfill = backfill.clone();
+                let in_progress = in_progress.clone();
+                let counter = processed_counter.clone();
+
+                async move {
+                    let start_cp = *range.start();
+                    let end_cp = *range.end();
+
+                    // Mark this chunk as in-progress
+                    in_progress.lock().await.insert(start_cp);
+
+                    // Execute backfill for the range
+                    if let Err(e) = backfill.backfill_range(pool, &range).await {
+                        let min_range_restart_num = {
+                            let mut guard = in_progress.lock().await;
+                            guard.remove(&start_cp);
+                            guard.iter().next().cloned().unwrap_or(start_cp)
+                        };
+                        error!("Chunk {start_cp}-{end_cp} failed. Minimum range restart number: {min_range_restart_num}. Error: {e}", );
```

Comment:
These futures are run concurrently, so `min_range_restart_num` might very well be the start of the last range of the concurrent batch.

What if no other ranges have been processed until this fails? Wouldn't the user be hinter toward restarting from a range that does not include the unprocessed ranges?

Reply:
Good catch! It indeed might be not the accurate restart number as it will very probably have gaps on failure.
Returning the exact restart number would require at least the book keeping of all currently running chunks.
This might be an optimization for the future.
In the worst case the restart point would be at `failing_start - max_concurrency * chunk_size`. An option would be to provide that entry point as part of the error message (this can be however also calculated by the user manually).

At this point I would suggest we remove the `Minimum range restart number: {min_range_restart_num}.` log as it's not accurate. And add a restart point (accurate or "by at most") in a follow up PR.

Reply:
Fixed as part of https://github.com/iotaledger/iota/pull/7603/commits/bb29d2e432fc4e702abc308df2552b6d135b6fd4